### PR TITLE
[path-balance-staked-and-locked] add new strategy for holders, stakers and locked tokens to vote

### PR DESCRIPTION
**Pull request by PathDAO**
Fixes # .

Changes proposed in this pull request:
- Add new strategy to calculate locked tokens in various contracts as well as current holders

### CODE REVIEW

File: src/strategies/path-balance-staked-and-locked/index.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import { BigNumberish } from '@ethersproject/bignumber';
+import { formatUnits } from '@ethersproject/units';
+import { Multicaller } from '../../utils';
+import { multicall } from '../../utils';
+
+export const author = 'bonustrack';
```
Comment (ChaituVR) [inline src/strategies/path-balance-staked-and-locked/index.ts:6]:
```suggestion
export const author = 'PathDAO';
```

File: src/strategies/path-balance-staked-and-locked/index.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+export const version = '0.1.1';
```
Comment (ChaituVR) [inline src/strategies/path-balance-staked-and-locked/index.ts:7]:
```suggestion
export const version = '0.1.0';
```

File: src/strategies/path-balance-staked-and-locked/schema.json [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        },
+        "address": {
+          "type": "string",
+          "title": "Contract address",
+          "examples": ["e.g. 0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984"]
+        },
+        "decimals": {
+          "type": "number",
+          "title": "Decimals",
+          "examples": ["e.g. 18"]
```
Comment (ChaituVR) [inline src/strategies/path-balance-staked-and-locked/schema.json:22]: There are other params missing in this? 🤔 Below are the params I see in examples.json
```
        "tokenAddress": "0x2a2550e0a75acec6d811ae3930732f7f3ad67588",
        "stakingAddress": "0x0bF6eeA205F46d796458D090F3aA333149287854",
        "lockedAddresses": [
          "0xDe48716A14C4CBc09656E21CE7F40FC1a02b3a25",
          "0x67f0260254FB3Cee97dA18077927888Ed72D1f17",
          "0xeE729DB66431e4401D63A38a2048d8CE0DF96eC3",
          "0x8923f2A0465287E2F8564F85a7f62ed34d947594",
          "0xfB05dc346f36eE88680e885CaC881983FAdEcBD4",
          "0x9A319F461A84EA5b508AE1e97113fD1743ED691D"
        ],
        "methodABI"
        ```

File: src/strategies/path-balance-staked-and-locked/index.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  for (let i = 0; i < callData.length; i++) {
+    const tempArray = await multicall(
+      network,
+      provider,
+      [options.methodABI],
+      callData[i],
+      { blockTag }
+    );
+    response.push(...tempArray);
+  }
```
Comment (ChaituVR) [inline src/strategies/path-balance-staked-and-locked/index.ts:65]: Something is not correct here. If the callData array is big ( in the case of a big number of voters) each address will send a separate request, which doesn't scale well. can you see if there is an alternate way to achieve this?

**Review by ChaituVR — CHANGES_REQUESTED** [review]

File: src/strategies/path-balance-staked-and-locked/index.ts (hunk 4 above)
Comment (PathDAO) [reply src/strategies/path-balance-staked-and-locked/index.ts:65]: thanks. let me try to modify this

File: src/strategies/path-balance-staked-and-locked/schema.json (hunk 3 above)
Comment (PathDAO) [reply src/strategies/path-balance-staked-and-locked/schema.json]: have added the missing params

File: src/strategies/path-balance-staked-and-locked/index.ts (hunk 4 above)
Comment (PathDAO) [reply src/strategies/path-balance-staked-and-locked/index.ts]: i think the new version should achieve better scaling by just using multicalls for the different contracts

File: src/strategies/path-balance-staked-and-locked/schema.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+          "type": "string",
+          "title": "Staking address",
+          "examples": ["e.g. 0x1f9840a85d5aF5bf1D1762F925BDADdC4201F984"]
+        },
+        "lockedAddresses": {
+          "type": "array",
+          "title": "Locked addresses",
+          "examples": ["e.g. [0x67f0260254FB3Cee97dA18077927888Ed72D1f17\", \"0xeE729DB66431e4401D63A38a2048d8CE0DF96eC3\",
+            \"0x8923f2A0465287E2F8564F85a7f62ed34d947594\",
+            \"0xfB05dc346f36eE88680e885CaC881983FAdEcBD4\"]"]
```
Comment (ChaituVR) [inline src/strategies/path-balance-staked-and-locked/schema.json:24]:
```suggestion
          "examples": ["e.g. [\"0x67f0260254FB3Cee97dA18077927888Ed72D1f17\", \"0xeE729DB66431e4401D63A38a2048d8CE0DF96eC3\", \"0x8923f2A0465287E2F8564F85a7f62ed34d947594\",\"0xfB05dc346f36eE88680e885CaC881983FAdEcBD4\"]"]
```

File: src/strategies/path-balance-staked-and-locked/index.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+export const author = 'PathDAO';
+export const version = '0.1.0';
+
+const tokenAbi = [
+  'function balanceOf(address account) external view returns (uint256)'
+];
+
+const stakingAbi = [
+  'function balanceOf(address account) external view returns (uint256)'
+];
```
Comment (bonustrack) [inline src/strategies/path-balance-staked-and-locked/index.ts:14]: These are duplicate code, can't this just be `const abi`?

Comment (PathDAO) [reply src/strategies/path-balance-staked-and-locked/index.ts:14]: yeah this will be fine let me just change that

**Comment by PathDAO** [comment]
@bonustrack sorry for pinging. Could you check this out and approve the changes if there aren't any issues? Would like to conduct a snapshot vote as soon as possible with our community. Thanks very much and appreciate the help!

**Review by bonustrack — APPROVED** [review]

**Review by ChaituVR — APPROVED** [review]

**Pull request merged**